cmd/thanos/store.go

```diff
@@ -49,42 +49,46 @@ import (
-	debugLogging                bool
-	syncInterval                time.Duration
-	blockSyncConcurrency        int
-	blockMetaFetchConcurrency   int
-	filterConf                  *store.FilterConfig
-	selectorRelabelConf         extflag.PathOrContent
-	advertiseCompatibilityLabel bool
-	consistencyDelay            commonmodel.Duration
-	ignoreDeletionMarksDelay    commonmodel.Duration
-	disableWeb                  bool
-	webConfig                   webConfig
-	postingOffsetsInMemSampling int
-	cachingBucketConfig         extflag.PathOrContent
-	reqLogConfig                *extflag.PathOrContent
-	lazyIndexReaderEnabled      bool
-	lazyIndexReaderIdleTimeout  time.Duration
+	indexCacheConfigs             extflag.PathOrContent
+	objStoreConfig                extflag.PathOrContent
+	dataDir                       string
+	disableCachingIndexHeaderFile bool
+	grpcConfig                    grpcConfig
+	httpConfig                    httpConfig
+	indexCacheSizeBytes           units.Base2Bytes
+	chunkPoolSize                 units.Base2Bytes
+	maxSampleCount                uint64
+	maxTouchedSeriesCount         uint64
+	maxConcurrency                int
+	component                     component.StoreAPI
+	debugLogging                  bool
+	syncInterval                  time.Duration
+	blockSyncConcurrency          int
+	blockMetaFetchConcurrency     int
+	filterConf                    *store.FilterConfig
+	selectorRelabelConf           extflag.PathOrContent
+	advertiseCompatibilityLabel   bool
+	consistencyDelay              commonmodel.Duration
+	ignoreDeletionMarksDelay      commonmodel.Duration
+	disableWeb                    bool
+	webConfig                     webConfig
+	postingOffsetsInMemSampling   int
+	cachingBucketConfig           extflag.PathOrContent
+	reqLogConfig                  *extflag.PathOrContent
+	lazyIndexReaderEnabled        bool
+	lazyIndexReaderIdleTimeout    time.Duration
 }
 
 func (sc *storeConfig) registerFlag(cmd extkingpin.FlagClause) {
 	sc.httpConfig = *sc.httpConfig.registerFlag(cmd)
 	sc.grpcConfig = *sc.grpcConfig.registerFlag(cmd)
 
-	cmd.Flag("data-dir", "Local data directory used for caching purposes (index-header, in-mem cache items and meta.jsons). If removed, no data will be lost, just store will have to rebuild the cache. NOTE: Putting raw blocks here will not cause the store to read them. For such use cases use Prometheus + sidecar.").
+	cmd.Flag("data-dir", "Local data directory used for caching purposes (index-header, in-mem cache items and meta.jsons). If removed, no data will be lost, just store will have to rebuild the cache. NOTE: Putting raw blocks here will not cause the store to read them. For such use cases use Prometheus + sidecar. Ignored if --disable-caching-index-header-file option is specified.").
 		Default("./data").StringVar(&sc.dataDir)
 
+	cmd.Flag("disable-caching-index-header-file", "Disable caching purposes (index-header, in-mem cache items and meta.jsons).").
```

Comment:
I think this needs to be explained better, and I also wonder how it will interact with `lazyIndexReaderEnabled=false`

Reply:
What about this:
>Disable caching index-header, and meta.jsons file on disk. If set, the store will ignore the `--data-dir`, and will not load the index header and meta.jsons files from disk at startup, and will not create the index header and meta.jsons files when synchronizing blocks.

This parameter just disables loading and creating index header and meta.jsons files, all data stored in memory is still the same as before, so it can interact with the --store.enable-index-header-lazy-reader parameter.

Reply:
Interesting, so the lazy reader does not require data to be on disk?

Reply:
The current logic is, the lazy reader will load the index header from the disk if the index header file exists, or build it from the index in the object store and write it on the disk when the first calling `load`, then the files on disk will never be used unless the block is deleted or store restarted.  so the lazy reader actually does not require the index header to be on the disk.

Reply:
Understood, thanks for the explanation 👍